# A plotting R script produced by the Revigo server at http://revigo.irb.hr/
# If you found Revigo useful in your work, please cite the following reference:
# Supek F et al. "REVIGO summarizes and visualizes long lists of Gene Ontology
# terms" PLoS ONE 2011. doi:10.1371/journal.pone.0021800

# --------------------------------------------------------------------------
# If you don't have the ggplot2 package installed, uncomment the following line:
# install.packages( "ggplot2" );
library( ggplot2 );

# --------------------------------------------------------------------------
# If you don't have the scales package installed, uncomment the following line:
# install.packages( "scales" );
library( scales );

# --------------------------------------------------------------------------
# Here is your data from Revigo. Scroll down for plot configuration options.

revigo.names <- c("term_ID","description","frequency","plot_X","plot_Y","log_size","value","uniqueness","dispensability");
revigo.data <- rbind(c("GO:0005975","carbohydrate metabolic process",5.521546466478484,1.7426823150854915,3.288646121635822,6.220115151808519,-6.958599418650251,0.9114241715060721,0.09441481),
c("GO:0005976","polysaccharide metabolic process",1.1167319130275588,0.19654886720154652,-7.352288282115867,5.526004379634945,-9.853899884150174,0.8902447103884149,0.15348198),
c("GO:0006091","generation of precursor metabolites and energy",2.62835465282124,6.268364249869623,-6.454897020565688,5.897738674144431,-16.356499948518014,0.9005256491863491,0.09910299),
c("GO:0006260","DNA replication",1.444034747678592,1.9858636593014038,-6.45260342727356,5.637632802972698,-6.58499992703721,0.8126289534108888,0.36026054),
c("GO:0006396","RNA processing",4.055292831628201,2.5767191523462323,-5.874624295514064,6.086076735055201,-18.657599034444775,0.7716244832761725,0.10860815),
c("GO:0006457","protein folding",1.0590625918025878,-5.23907883759791,-4.05484323606974,5.502977151016975,-7.853899884150174,0.9868337331400653,0.01196249),
c("GO:0006629","lipid metabolic process",4.558650411488468,5.779934623198953,5.61235094503985,6.136890764559154,-22.63830048845921,0.9144660334380033,0.0647278),
c("GO:0006725","cellular aromatic compound metabolic process",26.410310601416047,6.273240984001411,-3.807538397326007,6.8998277223761,-28.050600234038182,0.8405411765380995,0.15865379),
c("GO:0006793","phosphorus metabolic process",13.171532602957267,5.250047796087056,-5.239876734278923,6.597690579747489,-3.6990004060138233,0.8653653256128783,0.21430308),
c("GO:0006807","nitrogen compound metabolic process",48.837649518475466,8.16056483368987,-0.7614711637103984,7.1668089422978785,-31.065500033778324,0.8891574837708157,0.14874469),
c("GO:0006950","response to stress",4.90663876400189,-4.816530602207978,3.1868046361574103,6.168838538858752,-63.63830048845921,0.9600916624782455,0),
c("GO:0007049","cell cycle",1.795399173617054,-2.914544796473459,-6.85097297093843,5.732215984404404,-13.508600482164454,0.9858300236893759,0.01278857),
c("GO:0007154","cell communication",8.661893464294382,-0.05767719375625109,1.1036492576602104,6.415667163528361,-12.657599034444775,0.9816100704793129,0.02033788),
c("GO:0008150","biological_process",100,-0.9903813804219616,3.5205881462219977,7.47805417331953,-46.161200006196665,1,-0),
c("GO:0008152","metabolic process",65.3565173451011,-2.08181295740035,-3.246617616690469,7.293343082853126,-15.920800658725943,1,-0),
c("GO:0009056","catabolic process",6.059177012419161,-1.415491709060159,-7.742620070112594,6.260468037413432,-17.136699747577985,0.9366774573628209,0.07243865),
c("GO:0009058","biosynthetic process",28.23839282673206,8.473906996538163,1.635493214455846,6.928894185470589,-48.7958990181433,0.9078151391277827,0.06770307),
c("GO:0009725","response to hormone",0.46845263944413934,-4.615676344197652,3.4960560688049482,5.148722932785368,-21.853899884150174,0.9687720189972953,0.39089762),
c("GO:0009987","cellular process",79.12519251933975,-5.952774544606978,-2.2262295050755054,7.376368956838209,-64.60209994825834,1,-0),
c("GO:0010467","gene expression",9.706370753664654,0.7833735102598283,-5.142065787815958,6.465111183714701,-12.187100006469867,0.8758765857630529,0.21341582),
c("GO:0022402","cell cycle process",0.9753891897530218,2.7444807345943287,6.715013993547494,5.4672335778545404,-10.086199918147438,0.9869772662876957,0.01184321),
c("GO:0022900","electron transport chain",1.6021713573797571,1.25312995183406,-2.142821256437213,5.682764023912308,-4.431799449703461,0.9078256065049547,0.09277939),
c("GO:0032787","monocarboxylic acid metabolic process",2.668208950140285,4.583572272310628,-7.3123783178546935,5.904274536880065,-30.657599034444775,0.8344773523758396,-0),
c("GO:0033043","regulation of organelle organization",0.41984050830427583,-0.2155705932620626,6.459649756426299,5.101141938716288,-3.853899884150174,0.96393559529402,0.24057549),
c("GO:0034641","cellular nitrogen compound metabolic process",30.187049102942364,5.559019886407725,-3.1396006596476207,6.957874867913695,-24.49490023988286,0.8243711795726577,0.28457639),
c("GO:0042592","homeostatic process",0.6828582482895448,-5.804085967270446,1.3607414442046117,5.3123868339515,-4.853899884150175,1,-0),
c("GO:0043170","macromolecule metabolic process",38.34861513800723,5.925444105071126,0.6682875362510599,7.061803881678349,-11,0.8719684293875862,0.19244037),
c("GO:0044237","cellular metabolic process",49.22533251163365,7.887251712572975,0.1947824058484083,7.170242846650484,-29.16750002874279,0.8810204489321598,0.20673508),
c("GO:0044238","primary metabolic process",54.063119334849006,7.443646278630124,-0.7738689479721319,7.210955285610252,-14.920800658725943,0.8848213860327191,0.25761802),
c("GO:0044255","cellular lipid metabolic process",3.2652883239880963,7.848285759184705,-4.737001791438772,5.9919761366451345,-29.275700367766344,0.8871501357871294,0.10215517),
c("GO:0044281","small molecule metabolic process",15.15473791824727,7.235869079794617,4.343378374142536,6.658602684396198,-40.4684993687765,0.9223336306438839,0.09351135),
c("GO:0044703","multi-organism reproductive process",0.012745924497294354,-4.218132644199089,-2.509268396500507,3.5835388192543522,-3.102399846797664,1,-0),
c("GO:0046483","heterocycle metabolic process",26.396317359150885,5.6407703425991595,-3.9447958399362966,6.899597554781196,-27.744700955489275,0.8405637181533598,0.27023854),
c("GO:0050896","response to stimulus",14.674014998147983,-4.320003993376268,-5.415119239719931,6.644603215942828,-73.02689986820943,1,-0),
c("GO:0051128","regulation of cellular component organization",0.7940757528344382,0.36825019760805905,6.083511854803639,5.377917913026385,-4.619799615786161,0.9611770743975994,0.25809879),
c("GO:0051179","localization",18.792076186036212,-2.4656137988355966,-5.340397204327016,6.75202900021518,-7.167500028742789,1,-0),
c("GO:0051276","chromosome organization",1.2297688384231882,4.276740513898344,6.455376928260568,5.5678788176138525,-15.055500055117337,0.9189698232830183,0.01218534),
c("GO:0051641","cellular localization",3.7434567372987795,-2.9560176236009417,5.2474796781982995,6.051327363159859,-17.58499992703721,0.9205182613559715,0.01414867),
c("GO:0051649","establishment of localization in cell",2.1876556652574304,-2.678809384214302,5.487762439741995,5.818033784726208,-17.795899018143302,0.9164268077886867,0.38201368),
c("GO:0055082","intracellular chemical homeostasis",0.3917808477982281,-6.107184364875199,-0.4409510408437672,5.07110104767312,-3.7212006862444444,1,-0),
c("GO:0065007","biological regulation",23.052697479340427,-4.327863948554836,-0.035537394262861066,6.8407759726007304,-53.95859941865025,1,-0),
c("GO:0065008","regulation of biological quality",1.9965832141230166,0.5915156221591765,6.624491563493681,5.778342297928303,-18.677799318285587,0.9583119032951725,-0),
c("GO:0071554","cell wall organization or biogenesis",1.217525167213757,-2.784286915981528,1.802821214793683,5.5635332925219245,-11.61979961578616,0.9865830208577839,0.01217016),
c("GO:0071704","organic substance metabolic process",58.926970110141816,7.097845675577781,-0.20525440588222066,7.248368294629724,-19.63830048845921,0.8809139685551155,0.23810488),
c("GO:0071840","cellular component organization or biogenesis",9.440442605554349,-1.7804343645587335,-0.8096463379494037,6.453046668121093,-88.12489994136742,0.9813024933819848,0.01731557),
c("GO:0072524","pyridine-containing compound metabolic process",0.6882566395888947,2.256467217344018,-7.666101278677452,5.31580666742101,-10.036199899288166,0.8495042350405909,0.23958254),
c("GO:1901135","carbohydrate derivative metabolic process",6.41604626743983,3.83652142985249,4.564811177141776,6.285321871094186,-26,0.9178513841676051,0.07446158),
c("GO:1901360","organic cyclic compound metabolic process",27.27731619259278,6.314310583757705,2.004294905888208,6.9138558498071845,-30.420200403354908,0.8845186986659151,0.16673533),
c("GO:1901564","organonitrogen compound metabolic process",29.993834591427927,4.7728987137057866,-0.9273801561421829,6.955086199151694,-21.49490023988286,0.8500247670713024,0.30530756),
c("GO:1901576","organic substance biosynthetic process",27.07678740968754,4.898841717651884,1.3870415066627297,6.910651347384342,-35.619799615786164,0.7928693859089684,0.11462025));

one.data <- data.frame(revigo.data);
names(one.data) <- revigo.names;
one.data <- one.data [(one.data$plot_X != "null" & one.data$plot_Y != "null"), ];
one.data$plot_X <- as.numeric( as.character(one.data$plot_X) );
one.data$plot_Y <- as.numeric( as.character(one.data$plot_Y) );
one.data$log_size <- as.numeric( as.character(one.data$log_size) );
one.data$value <- as.numeric( as.character(one.data$value) );
one.data$frequency <- as.numeric( as.character(one.data$frequency) );
one.data$uniqueness <- as.numeric( as.character(one.data$uniqueness) );
one.data$dispensability <- as.numeric( as.character(one.data$dispensability) );
#head(one.data);


# --------------------------------------------------------------------------
# Names of the axes, sizes of the numbers and letters, names of the columns,
# etc. can be changed below

p1 <- ggplot( data = one.data );
p1 <- p1 + geom_point( aes( plot_X, plot_Y, colour = value, size = log_size), alpha = I(0.6) );
p1 <- p1 + scale_colour_gradientn( colours = c("blue", "green", "yellow", "red"), limits = c( min(one.data$value), 0) );
p1 <- p1 + geom_point( aes(plot_X, plot_Y, size = log_size), shape = 21, fill = "transparent", colour = I (alpha ("black", 0.6) ));
p1 <- p1 + scale_size( range=c(5, 30)) + theme_bw(); # + scale_fill_gradientn(colours = heat_hcl(7), limits = c(-300, 0) );
ex <- one.data [ one.data$dispensability < 0.15, ];
p1 <- p1 + geom_text( data = ex, aes(plot_X, plot_Y, label = description), colour = I(alpha("black", 0.85)), size = 3 );
p1 <- p1 + labs (y = "semantic space x", x = "semantic space y");
p1 <- p1 + theme(legend.key = element_blank()) ;
one.x_range = max(one.data$plot_X) - min(one.data$plot_X);
one.y_range = max(one.data$plot_Y) - min(one.data$plot_Y);
p1 <- p1 + xlim(min(one.data$plot_X)-one.x_range/10,max(one.data$plot_X)+one.x_range/10);
p1 <- p1 + ylim(min(one.data$plot_Y)-one.y_range/10,max(one.data$plot_Y)+one.y_range/10);


# --------------------------------------------------------------------------
# Output the plot to screen

p1;

# Uncomment the line below to also save the plot to a file.
# The file type depends on the extension (default=pdf).

# ggsave("/path_to_your_file/revigo-plot.pdf");

